torch/utils/data/graph.py

```diff
@@ -41,18 +41,21 @@ def reduce_hook(obj):
             return stub_unpickler, ()
 
     # TODO(VitalyFedyunin):  Better do it as `with` context for safety
-    IterableDataset.set_reduce_ex_hook(reduce_hook)
+    IterDataPipe.set_reduce_ex_hook(reduce_hook)
```

Comment:
Would using the `try... finally` pattern help with safety? Or do we have to use `with`?

```python
try:
    IterDataPipe.set_reduce_ex_hook(reduce_hook)
    if exclude_primitive:
        IterDataPipe.set_getstate_hook(getstate_hook)
    p.dump(scan_obj)
except AttributeError:  # unpickable DataPipesGraph
    pass  # TODO(VitalyFedyunin): We need to tight this requirement after migrating from old DataLoader
finally:
    IterDataPipe.set_reduce_ex_hook(None)
    if exclude_primitive:
        IterDataPipe.set_getstate_hook(None)
```